File: src/main/java/com/sashwat/java/O4_semantic_validations/ValidationsDemo.java
package com.sashwat.java.O4_semantic_validations;

import com.google.common.base.Preconditions;
import org.apache.commons.collections4.CollectionUtils;
import org.apache.commons.lang3.Validate;

import java.util.List;

public class ValidationsDemo {

    public String join(String input, List<String> values) {

        Validate.notBlank(input, "");
        Preconditions.checkArgument(input.length() >= 2, "Input length should be atleast 2");

        Preconditions.checkArgument(CollectionUtils.isNotEmpty(values), "Values cannot be empty");


        Preconditions.checkState(true);
        // ...

        return "";
    }
}

File: src/main/kotlin/com/sashwat/kotlin/O4_semantic_validations/ValidationsDemo.kt
package com.sashwat.kotlin.O4_semantic_validations

class ValidationsDemo {

    fun join(input: String?, values: List<String>) {
        requireNotNull(input) {"Input cannot be null"}

        require(input.length >= 2) {"Input length should be atleast 2"}
        require(values.isNotEmpty()) {"Values cannot be null"}

        check(true) { "State condition failed"}
    }
}